akka-docs/src/test/java/jdocs/tutorial_5/DeviceGroup.java

```diff
@@ -139,8 +139,14 @@ private void onTerminated(Terminated t) {
   //#query-added
 
   private void onAllTemperatures(RequestAllTemperatures r) {
+    // since Java collections are mutable, we want to avoid sharing them between actors (since multiple Actors (threads)
+    // modifying the same mutable data-structure is not safe), and perform a defensive copy of the mutable map:
+    //
+    // Feel free to use your favourite immutable data-structures library with Akka in Java applications!
+    HashMap<ActorRef, String> copyActorToDeviceId = new HashMap<>(this.actorToDeviceId);
+
     getContext().actorOf(DeviceGroupQuery.props(
-            actorToDeviceId, r.requestId, getSender(), new FiniteDuration(3, TimeUnit.SECONDS)));
+        copyActorToDeviceId, r.requestId, getSender(), new FiniteDuration(3, TimeUnit.SECONDS)));
```

Comment:
Now it sounds like a method, wouldn't `actorToDeviceIdCopy` be better?

Reply:
Good point, thanks